### ISSUE TITLE
Review issuance function implementation

### ISSUE THREAD

**Issue by danlessa** [BlockScience/subspace#33]
A/C: write-up on conclusions

**Comment by danlessa** [comment]
Conclusions:

- Technically, the issuance during multiple blocks should be estimated, as the Rewards over a day is equal to the sum over rewards over all the blocks during that day. One approach is to us to define the reward to be issued during the timestep as follows: `rewards_during_timestep = block_reward[t] * blocks_per_timestep`, where `blocks_per_timestep = params['timestep_in_days'] * params['block_time_in_seconds'] * (24 * 60 * 60)

- `a` should be defined as `a = S_r - b * tanh(c * d)`
- `b` should be defined as `b = S_r - max(S_r - F_bar, 0) / tanh(c)`
- `F_bar = MAX_NORMAL_BLOCK_LENGTH * state['storage_fee_in_credits_per_bytes']`

The variable `storage_fee_in_credits_per_bytes` (defined at `p_storage_fees`) should be added as a new state variable

**Comment by LinuxIsCool** [comment]
@danlessa Can you clarify `MAX_NORMAL_BLOCK_LENGTH`? It's not yet in the codebase.

**Comment by LinuxIsCool** [comment]
@danlessa one more question about the type of `storage_fee_in_credits_per_bytes`. I guess it should be in Credits / Bytes. I'm not sure how to achieve this as a type annotation. Are there units in https://github.com/BlockScience/subspace/blob/main/subspace_model/types.py that I can use for this case?

**Comment by LinuxIsCool** [comment]
I believe we already had this state property as `transaction_byte_fee`. I'm assuming they are the same concept, and I have refactored `transaction_byte_fee` to `storage_fee_in_credits_per_bytes`. Let me know if this sounds correct.

**Comment by LinuxIsCool** [comment]
I'll review this reference as I continue: https://subspacelabs.notion.site/Dynamic-Issuance-Specification-3bdf63955cb943489347889775d71c24

**Comment by danlessa** [comment]
> @danlessa Can you clarify `MAX_NORMAL_BLOCK_LENGTH`? It's not yet in the codebase.

that's `params['max_block_size']`

> @danlessa one more question about the type of `storage_fee_in_credits_per_bytes`. I guess it should be in Credits / Bytes. I'm not sure how to achieve this as a type annotation. Are there units in https://github.com/BlockScience/subspace/blob/main/subspace_model/types.py that I can use for this case?

Create a new unit called `CreditsPerBytes` or `ShannonsPerByte`

> I believe we already had this state property as `transaction_byte_fee`. I'm assuming they are the same concept, and I have refactored `transaction_byte_fee` to `storage_fee_in_credits_per_bytes`. Let me know if this sounds correct.

@LinuxIsCool the Transaction Fee is defined as Storage Fee + Compute Fee + Priority Fee. The Issuance Mechanisms uses the Cummulative Storage Fee in the Proposed Block (eg. sum over storage fees for all txs)

**Comment by LinuxIsCool** [comment]
Blocked by #45 

Because system is raising ValueError now that history_size > space_pledged.

**Comment by LinuxIsCool** [comment]
> Create a new unit called `CreditsPerBytes` or `ShannonsPerByte`
> 
> > I believe we already had this state property as `transaction_byte_fee`. I'm assuming they are the same concept, and I have refactored `transaction_byte_fee` to `storage_fee_in_credits_per_bytes`. Let me know if this sounds correct.
> 
> @LinuxIsCool the Transaction Fee is defined as Storage Fee + Compute Fee + Priority Fee. The Issuance Mechanisms uses the Cummulative Storage Fee in the Proposed Block (eg. sum over storage fees for all txs)

@danlessa it looks like `transaction_byte_fee` is in units of CreditsPerByte and represents the cost of storage per byte. 

    transaction_byte_fee: CreditsPerByte = total_credit_supply / free_space
    
And that the sum of storage fees for all transactions as you describe would be in units of Credits instead of CreditsPerByte as the following: 

    storage_fee_volume: Credits = transaction_byte_fee * extrinsic_length_in_bytes

We can discuss the above in our session today.

**Comment by LinuxIsCool** [comment]
Refactor:

`transaction_byte_fee` to `storage_fee_in_credits_per_bytes`

And follow up review bundle behavior in regards to storage fees in another issue.

**Comment by LinuxIsCool** [comment]
I've re-opened, I just realized, I did the preliminary work, but hadn't actually updated the issuance function. I'm updating now and will also post a write-up on conclusions to get the A/C.

**Comment by LinuxIsCool** [comment]
Update has been made here:  https://github.com/BlockScience/subspace/commit/7b80fc6ecb919c664fe4e878d6aed08f764ac9c7

Sanity Check Notebook can be seen here: https://github.com/BlockScience/subspace/blob/22072029ce0bb4b0523ec0b0189e272d0e344be3/notebooks/01-sanity_check_run.ipynb

Images can be found here: https://github.com/BlockScience/subspace/tree/22072029ce0bb4b0523ec0b0189e272d0e344be3/notebooks/images
